### DISCUSSION TITLE
Heterogeneous graphs message passing

### DISCUSSION THREAD

**Discussion by Petto11** [pyg-team/pytorch_geometric#9156] [Q&A]
Hi everyone
I am working on a prediction task on order fulfillment time, where I have 3 types of nodes in my HeteroGraph, which are order, item and event. This is the architecture I am currently using:
```python
class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels, num_layers):
        super().__init__()

        self.convs = torch.nn.ModuleList()
        for _ in range(num_layers):
            conv = HeteroConv({\
                ('event', 'to', 'event'): GCNConv(-1, hidden_channels),
                ('order', 'to', 'item'): SAGEConv((-1, -1), hidden_channels ),#,add_self_loops = False),
                ('order', 'to', 'event'): SAGEConv((-1, -1), hidden_channels, aggr = 'max'),#, add_self_loops = False),
                ('item', 'to', 'event'): SAGEConv((-1, -1), hidden_channels, aggr = 'max'),#, add_self_loops = False),
                ('item', 'rev_to', 'order'): SAGEConv((-1, -1), hidden_channels),#, add_self_loops = False),
                ('event', 'rev_to', 'order'): SAGEConv((-1, -1), hidden_channels + 1),#, add_self_loops = False),
                ('event', 'rev_to', 'item'): SAGEConv((-1, -1), hidden_channels)#, add_self_loops = False)
            }, aggr='sum')
            self.convs.append(conv)

        self.lin = Linear(hidden_channels, out_channels)

    def forward(self, x_dict, edge_index_dict):
        for conv in self.convs:
            x_dict = conv(x_dict, edge_index_dict)
            x_dict = {key: F.relu(x) for key, x in x_dict.items()}
        return self.lin(x_dict['order']) + self.lin(x_dict['event'][-1]), x_dict['event']


model = HeteroGNN(hidden_channels=5, out_channels=1,
                 num_layers=3
```
From what I understood, when I use SAGEConv in HeteroGraphs I update the node embedding of the source node with its embedding + the aggregation of the neighbor embeddings (multiplied by W). The problem is that when I comment all the layers having "event" as a source node I still get the embedding of the node 'event' updated, while If I comment all the nodes with event as a target node, I get the following message:
```
There exist node types ({'event'}) whose representations do not get updated during message passing as they do not occur as destination type in any edge type
```
For me this is counterintuitive since if a node is not in any target, it will still get his embedding updated because it has neighbors. Can you please help me figure out why?

**Comment by rusty1s** [comment]
> The problem is that when I comment all the layers having "event" as a source node I still get the embedding of the node 'event' updated.

Can you clarify your concern? Message passing sends information from source nodes to destination nodes. Every edge type in a heterogeneous graph thus updates the destination node representation.
